lib/puppet/node.rb

```diff
@@ -108,12 +108,29 @@ def fact_merge
   # Merge any random parameters into our parameter list.
   def merge(params)
     params.each do |name, value|
-      @parameters[name] = value unless @parameters.include?(name)
+      if @parameters.include?(name)
+        Puppet::Util::Warnings.warnonce("The node paramters #{name} for node #{@parameters["hostname"]} was already set to #{@parameters[name]}. It could not be set to #{value}")
+      else
+        @parameters[name] = value
+      end
     end
 
     @parameters["environment"] ||= self.environment.name.to_s
   end
 
+  def add_server_facts(server_facts)
+   # Append the current environment to the list of server facts
+    complete_server_facts = server_facts.merge({ "environment" => self.environment.name.to_s})
+
+   # Set the top scope variable $server_facts if :trusted_server_facts is true
+   if Puppet[:trusted_server_facts]
+     @topscope.set_trusted(complete_server_facts)
+   end
+
+    # Merge the server facts into the parameters for the node
+    merge(server_facts)
```

Comment:
You can leave it the way it was. I was concerned that it may warn about overwrite for environment, but it depends on how many time `merge` is called; it may only be done once. OTOH, it may be expected that the parameter for environment is set after a merge. As long as it is not set more than once it should be fine (and if set more than once, there should be no warning about an "overwrite"